### DISCUSSION TITLE
[New] Favoriser l'utilisation des PWA vs les applications natives et hybrides

### DISCUSSION THREAD

**Discussion by florinesueur** [cnumr/best-practices#122] [Bonnes pratiques] [answered]
Si cela correspond au projet, les avantages des PWA sont :

- les coûts de développement et de maintenance sont limité
- plus légères, moins consommatrices de ressources et de bande passante
- fonctionne en hors connexion
- mise en cache
- mise à jour facile
- accessible aussi sur desktop

Sources :
- [https://www.researchgate.net/publication/340389357_Evaluating_the_Impact_of_Caching_on_the_Energy_Consumption_and_Performance_of_Progressive_Web_Apps](https://www.researchgate.net/publication/340389357_Evaluating_the_Impact_of_Caching_on_the_Energy_Consumption_and_Performance_of_Progressive_Web_Apps)
- [https://www.researchgate.net/publication/330834334_Impact_of_Progressive_Web_Apps_on_Web_App_Development](https://www.researchgate.net/publication/330834334_Impact_of_Progressive_Web_Apps_on_Web_App_Development)

**Comment by ACTLEM** [comment]
On touche plus les applications mobiles qu'un service web même si sur le fond je suis d'accord...je penche plus pour un futur référentiel d'écoconception d'applications mobiles...mais à l'écoute d'autres avis.

**Reply by FredMos28** [reply]
La PWA est plutôt un modèle hybride entre l'application Web et l'application mobile.  L'idée étant de proposer la même expérience à l'utilisateur sur desktop et sur mobile.

**Reply by tbroyer** [reply]
Une PWA est une application web "on steroids" qui rend de fait les applications natives (mobiles ou non) redondantes.

* https://excalidraw.com/
* https://photoshop.adobe.com (pas sûr qu'on puisse vraiment l'appeler une PWA, je n'en sais pas assez encore)
* forbes.com
* les générateurs d'attestation de déplacement pendant les confinements
* AliExpress, Twitter, etc. (les applis mobiles existent –essentiellement pour la visibilité dans les stores, et probablement beaucoup aussi historiquement–, mais elles sont redondantes)

**Reply by brpasi** [reply]
Dans le fond, ça reste effectivement une appli web (qui fait office d'app mobile au passage), donc plutôt d'accord aussi que cette BP a sa place.
Peut être à intégrer à la BP # 015 Choisir les technologies les plus adaptées ?

**Reply by ACTLEM** [reply]
Merci pour ces avis. J'ai toutefois toujours quelques interrogations sur cette bonne pratique:

- D'une manière générale, j'ai du mal à voir l'avantage d'une PWA vs un site web classique en termes d'impacts
- Firefox semble avoir fait machine arrière sur le support des PWA, c'est un point qui me dérange
- Comme le signale @brpasi, cela rentre plus dans la BP #15

NB: Face à une application mobile native, le fait d'être non lié à un OS, la réduction du risque d'obsolescence, la PWA a par contre un intérêt.

**Reply by brpasi** [reply]
- D'une manière générale, j'ai du mal à voir l'avantage d'une PWA vs un site web classique en termes d'impacts
> vs un site web classique : pas de différence. 
> vs un site web classique ET une app mobile : tu économises le dév d'une app mobile, qui sera elle même bien plus "lourde" que la version PWA. 
> La préco est donc ici de se dire que si le site web doit appeler une déclinaison mobile, autant penser en PWA 
- Firefox semble avoir fait machine arrière sur le support des PWA, c'est un point qui me dérange
> La version web "de base" n'en reste pas moins compatible. Ce risque, c'est toutefois le lot de toutes les technos dans l'absolu. D'ou l'idée de plutôt raccrocher ce point à titre d'exemple uniquement à la BP # 015 ;)

**Reply by ACTLEM** [reply]
Oui, nous sommes en phase pour la BP #15 :-) 
De plus , comme c'est un référentiel de recommandations d'écoconception web (surtout sites web) et non d'applications mobiles (les leviers de réduction d'impacts ne sont pas strictement identiques et les bonnes pratiques peuvent différer) avoir une BP spécifique ayant des réductions d'impacts qu'en cas d'application mobile ne me semble pas adéquat.

**Comment by tbroyer** [comment]
À noter qu'une PWA n'est pas forcément utilisable hors ligne (Twitter hors ligne, ça n'a pas vraiment de sens)

Et concernant Twitter, ils développent avec React Native et réutilisent donc une très grande partie du code entre les applications mobiles natives et l'application Web (via React Native for Web), et ça se voit : https://giuseppegurgone.com/twitter-html/
Ça reste tout de même bien moins pire que Flutter par exemple (cf. https://github.com/cnumr/best-practices/discussions/28#discussioncomment-1528417)

Ayant utilisé l'application Twitter native il y a quelques années et ayant basculé sur la PWA depuis longtemps (historiquement mobile.twitter.com, mais qui est devenue depuis le "nouveau Twitter" sur twitter.com, et qui est installable en "application desktop" depuis le Windows Store), je ne comprend pas qu'ils continuent à maintenir ces applications natives, à part peut-être pour conserver leurs (très nombreux) utilisateurs mobiles (bon, il y a quelques fonctionnalités propres aux applis natives il me semble, comme d'ajouter des masques et emojis sur les images avant de les partager, mais rien a priori qui n'aurait pu être fait dans la PWA non plus s'ils avaient voulu s'en donner les moyens).

Donc pour moi :+1: pour cette BP

**Reply by EmmanuelDemey** [reply]
Pour moi le plus important dans PWA c'est le P de progressive (qui devrait faire office d'une règle a elle même). Donc ce sera fonctionnel que pour les navigateurs les supportant, sans dégrader les autres. 

Une PWA n'est pas juste une site web installable, mais c'est surtout le SW. 

Quand on dit que Firefox ne supporte plus les PWA, il ne supporte plus le côté instabilité.en fait. Le SW est toujours fonctionnel (de toute façon c'est standardisé)

Un SW est fonctionnel même si l'application n'est pas installée. Il est également utilisé lors d'une navigation normal via le navigateur. 

Pour les SW, j'ajouterai que les développeurs doivent faire attention aux stratégies de mise en cache implémentée. Car si mal implémentée, cela peut engendrer une mauvaise UX. 

Dans cette famille de bonnes pratiques, on pourrait lister : 
- la définition de l'appshell
- utilisation de IndexedDB

**Reply by ACTLEM** [reply]
Merci Emmanuel. Finalement, parler PWA permet d'insister sur l'amélioration continue (progressive) qui est un des fondamentaux de la démarche d'écoconception. En plus d'être un exemple de la BP 15, cela pourrait aussi alimenter la page hors bonnes pratiques sur la démarche d'écoconception (voir Issue https://github.com/cnumr/best-practices/issues/139).
Je reste toujours peu convaincu que cela doive faire l'objet d'une BP en soit, mais je ne demande qu'à l'être :-)
Peut-être avoir une BP plus spécifique aux SW, mais sur le coup j'ai du mal à voir ce que cela pourrait apporter en termes d'écoconception.

**Reply by EmmanuelDemey** [reply]
Un SW peut permettre de limiter le nombre de requête HTTP en fonction de la "freshiness" de la donnée. 
Ça fait office de cache. Mais dont tu as la main programmatiquemment, contrairement au cache HTTP

**Reply by tbroyer** [reply]
Pour le SW, cf. #120 

Le seul problème que j'ai avec le terme "PWA" c'est que ce n'est pas _une_ technologie mais un ensemble de technologies + de pratiques (progressive)
Concernant les technos:
* web manifest pour _l'installabilité_
* service worker pour l'interception des requêtes HTTP (ce qui permet aussi un usage hors-ligne), cf #120 
* l'API de cache pour… la mise en cache de réponses HTTP (souvent utilisé pour l'usage hors ligne, mais pas que), et là il pourrait être intéressant d'avoir une BP pour éviter les mauvais usages; cf. https://github.com/cnumr/best-practices/discussions/109#discussioncomment-1545497

Mais je pense que ça pourrait être intéressant d'avoir une BP "web vs native" pour éviter les "doublons" (et même si on peut "développer une fois, générer native + web", les exemples Flutter (cf. https://github.com/cnumr/best-practices/discussions/28#discussioncomment-1528417 à nouveau) sont plutôt rédhibitoire), comparaison "web vs native" rendue possible par les PWA.

(note: l'appli native Twitter pour Android semble faire sensiblement le même poids que le cache de la PWA ! …mais c'est parce que Twitter décide de précharger la totalité de l'application : environ 32Mo)

**Reply by ACTLEM** [reply]
Effectivement, la #120 répond déjà au service worker. Merci @tbroyer.
Pour une BP "web vs native", je fais continuer à faire le rabat joie :-D...il s'agit de bonnes pratiques écoconception web donc le choix a déjà été fait. Par contre, c'est un argument de plus pour créer une liste de bonnes pratiques pour les applications mobiles ;-)

**Reply by ldesgrange** [reply]
Un BP "web vs native" m'intéresse, personnellement n'ayant pas fait d'app mobiles, je ne vois pas bien comment une PWA bien faite peut-être plus écoresponsable qu'une app native bien faite.

**Comment by ACTLEM** [comment]
@florinesueur, c'est possible de créer un point visio (15-30mins) pour discuter avec les personnes qui seraient intéressées afin de proposer concrètement une bonne pratique **web** sachant qu'il y en a déjà une autour des services workers stp?

**Reply by EmmanuelDemey** [reply]
Pas avant la semaine prochaine. Je suis en congés.

**Reply by florinesueur** [reply]
@tbroyer @EmmanuelDemey  @ACTLEM @brpasi @FredMos28 et tous ceux qui souhaitent discuter de cette proposition dans un point visio de 15-30 minutes, merci d'indiquer vos disponibilités [https://framadate.org/HV9ElAhLZ58HDsIi](https://framadate.org/HV9ElAhLZ58HDsIi) au plus tard avant le lundi 8 novembre.

**Comment by florinesueur** [comment]
Merci à @tbroyer @ACTLEM @brpasi d’avoir répondu au framadate. Souhaitez-vous décaler à la semaine du 15 pour être plus nombreux notamment avec @EmmanuelDemey @FredMos28 et @devcrafting ? Si c'est le cas merci de me dire quand vous êtes plutôt disponible matin, midi, fin d'après-midi ou plutôt en soirée.

**Reply by ACTLEM** [reply]
Ok pour moi pour décaler la semaine du 15 (plutôt disponible le midi ou en soirée sauf lundi 15/11 midi).

**Reply by brpasi** [reply]
Plutôt le midi pour moi (sauf le mercredi) ou soirée le lundi

**Reply by EmmanuelDemey** [reply]
Dispo lundi, mardi et mercredi. Si c'est pour 15mn, dispo tout le temps :)

**Reply by devcrafting** [reply]
je ne suis pas sûr d'être suffisamment au fait sur les PWA pour bien argumenter, après si c'est 15/30min, c'est jouable le mardi 16/11 matin/midi si vous voulez un avis non expert :).

**Reply by florinesueur** [reply]
Voici le nouveau sondage : https://framadate.org/tETMZiycgnZm9yMb.
J'ai mis 30min, mais je pense qu'en 15 min, c'est jouable.

**Reply by florinesueur** [reply]
@tbroyer @ACTLEM @brpasi @EmmanuelDemey @FredMos28 @devcrafting et tous, rendez-vous demain **lundi 15 nov. à 18h**. 

Lien de la réunion: [kmeet.infomaniak.com/k4wkxzsq9n6korqfzc3sdtgz](https://kmeet.infomaniak.com/k4wkxzsq9n6korqfzc3sdtgz)
Code de la réunion: 2105680346

**Comment by ACTLEM** [comment, marked as answer]
@florinesueur Je ne vois pas d'infos sur ce que nous nous sommes dits (Tourner la BP sur favoriser une PWA si applicative native liée au site web, éviter double maintenance)?

A noter que cette discussion est liée à la https://github.com/cnumr/best-practices/discussions/199

**Reply by florinesueur** [reply]
En effet. J'étais encore en réflexion.

Proposition de titre : 
- Favoriser une PWA à une application mobile native associée à un site
- ...

Si une application mobile est strictement nécessaire et qu'elle reprend entièrement ou en partie le site internet alors une PWA est une solution évitant un double, voire un triple, développement et maintenance (permettant de réduire la dette technique).

Piste d'exemples que je vais fouiller :
- E-commerce comme Etam, ....
- SaaS comme Clickup, Notion, Miro, ...

**Reply by ACTLEM** [reply]
J'ai créé l'issue pour que tu puisses proposer quelque chose concrètement via une PR: https://github.com/cnumr/best-practices/issues/308
